src/phoenix/db/facilitator.py

```diff
@@ -0,0 +1,126 @@
+from __future__ import annotations
+
+import asyncio
+from functools import partial
+
+from sqlalchemy import (
+    distinct,
+    func,
+    insert,
+    select,
+    update,
+)
+from sqlalchemy.ext.asyncio import AsyncSession
+from sqlalchemy.sql.functions import coalesce
+
+from phoenix.auth import compute_password_hash
+from phoenix.config import ENABLE_AUTH, PHOENIX_SECRET
+from phoenix.db import models
+from phoenix.db.enums import COLUMN_ENUMS, AuthMethod, UserRole
+from phoenix.server.types import DbSessionFactory
+
+
+class Facilitator:
+    """
+    Facilitates the creation of database records necessary for Phoenix to function. This includes
+    ensuring that all enum values are present in their respective tables, ensuring that all user
+    roles are present, and ensuring that the admin user has a password hash. These tasks will be
+    carried out as callbacks at the very beginning of Starlette's lifespan process.
+    """
+
+    def __init__(self, db: DbSessionFactory) -> None:
+        self._db = db
+
+    async def __call__(self) -> None:
+        async with self._db() as session:
+            for fn in (
+                _ensure_enums,
+                _ensure_user_roles,
+                *((_ensure_admin_password,) if ENABLE_AUTH else ()),
+            ):
+                async with session.begin_nested():
+                    await fn(session)
+
+
+async def _ensure_enums(session: AsyncSession) -> None:
+    """
+    Ensure that all enum values are present in their respective tables. If any values are missing,
+    they will be added. If any values are present in the database but not in the enum, an error will
+    be raised. This function is idempotent: it will not add duplicate values to the database.
+    """
+    for column, enum in COLUMN_ENUMS.items():
```

Comment:
I'm a little confused, don't these roles get added with `_ensure_user_roles`?

Reply:
Feels a bit abstract given that we only have `UserRoles` right now.

Reply:
I think these enums are intended for look-up tables, so they should have no upstream dependencies